### PR TITLE
Port some CoreRT Threading classes to Mono

### LINKED ISSUE #44795: Move System.Threading.Event/Mutex/Semaphore support to managed code
These synchronization classes have little to do with the runtime and they can be implemented mostly in managed code as done in corert.

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Draft until I see how this behaves on CI.

Fixes https://github.com/dotnet/runtime/issues/44795

A few notes:

I've attempted to nullable annotate the relevant files but would appreciate someone else checking over my work, since the design does not really lend itself to that.

I can split the time-related parts of the System.Native addition off into a separate header if desired, but since they wouldn't be used elsewhere I just put the relevant includes in the C file.

This can fairly easily be moved to shared if desired, but I've put it in Mono only for now. Given there's a chance we might want to share it eventually after testing in Mono, I've left a CORERT define.

The `TARGET_UNIX` define usage isn't ideal, but given we care about the layout for that particular class it seemed like the easiest option. If it's a problem please let me know.

### CODE REVIEW

File: src/libraries/Native/Unix/System.Native/pal_threading.c [hunk 1]
```diff
@@ -119,6 +126,49 @@ void SystemNative_LowLevelMonitor_Wait(LowLevelMonitor* monitor)
     SetIsLocked(monitor, true);
 }
 
+uint32_t SystemNative_LowLevelMonitor_TimedWait(LowLevelMonitor* monitor, int32_t timeoutMilliseconds)
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/pal_threading.c:129]: The support for timed waits needs more changes in this file. It would be best if you can start with verbatim copy of https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/src/libraries/Native/Unix/System.Native/pal_threading.c

File: src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (jkotas) [inline src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:1]: Can you add these files to the shared partition, and ifdef them out for CoreCLR in the build script for now?

We will want to switch CoreCLR to this eventually as well, and having it in the shared place will makes it easier. It is similar how the portable threadpool was done. It was enabled for Mono only, and later for CoreCLR too.

File: src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Comment (jkotas) [inline src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:3]: Have you done any changes to these files? For review, I would be useful to have separate commit (or even PR) that just copies the files over, so that it easy to review any changes you have done on top of it.

Comment (CoffeeFlux) [reply src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:3]: I have indeed made changes, mostly the annotations but also removing some CoreRT-specific invocations. Happy to go back and split the commit up, it was just annoying to do initially since I was fixing a host of build errors.

File: src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs (hunk 2 above)
Comment (CoffeeFlux) [reply src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:1]: Will do.

File: src/mono/netcore/System.Private.CoreLib/src/System/Threading/EventWaitHandle.Unix.Mono.cs [hunk 4]
```diff
@@ -2,98 +2,82 @@
...
 
 using Microsoft.Win32.SafeHandles;
-using System.Runtime.CompilerServices;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
 
 namespace System.Threading
 {
     public partial class EventWaitHandle
```
Comment (jkotas) [inline src/mono/netcore/System.Private.CoreLib/src/System/Threading/EventWaitHandle.Unix.Mono.cs:11]: I do not see anything Mono-specific in this file. Can it be in the shared partition and not have .Mono in name?

File: src/libraries/Native/Unix/System.Native/pal_threading.c (hunk 1 above)
Comment (CoffeeFlux) [reply src/libraries/Native/Unix/System.Native/pal_threading.c:129]: I guess I was working off an older copy, and was also referencing the Mono implementation. I can just bring that and `pal_time.h` over verbatim if that's preferable.

File: src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs (hunk 3 above)
Comment (jkotas) [reply src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:3]: The most up to date version of these files is in https://github.com/dotnet/runtimelab/tree/feature/NativeAOT/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading. Not sure whether you have copied them from there, or somewhere else.

Comment (jkotas) [reply src/mono/netcore/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:3]: >  I would be useful to have separate commit (or even PR)

Looks like we may be iterating on this a bit, so I would recommend to submit PR that just copies the needed files over, without any changes and without hooking it to the build. And then separate PR to hook it up to the build and make the necessary adjustments.

**Comment by CoffeeFlux** [comment]
Splitting up into multiple PRs per feedback.
